// @bigcommerce/bigpay-client/lib/store/store-request-sender.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _requestSender = _interopRequireDefault(require("../common/http-request/request-sender"));

var _methodTypes = require("../common/http-request/method-types");

var _urlHelper = _interopRequireDefault(require("./url-helper"));

var _mappers = require("./v2/mappers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var StoreRequestSender = /*#__PURE__*/function () {
  _createClass(StoreRequestSender, null, [{
    key: "create",

    /**
     * @param {Object} config
     * @returns {StoreRequestSender}
     */
    value: function create(config) {
      var urlHelper = _urlHelper.default.create(config);

      var requestSender = _requestSender.default.create();

      return new StoreRequestSender(urlHelper, requestSender);
    }
    /**
     * @param {UrlHelper} urlHelper
     * @param {RequestSender} requestSender
     * @returns {void}
     */

  }]);

  function StoreRequestSender(urlHelper, requestSender) {
    _classCallCheck(this, StoreRequestSender);

    /**
     * @private
     * @type {UrlHelper}
     */
    this.urlHelper = urlHelper;
    /**
     * @private
     * @type {RequestSender}
     */

    this.requestSender = requestSender;
  }
  /**
   * @param {Object} data
   * @param {Function} [callback]
   * @return {void}
   */


  _createClass(StoreRequestSender, [{
    key: "loadInstruments",
    value: function loadInstruments(data, callback) {
      var url = this.urlHelper.getInstrumentsUrl(data.storeId, data.customerId, data.currencyCode);
      var options = {
        headers: (0, _mappers.mapToHeaders)(data)
      };
      this.requestSender.sendRequest(url, null, options, callback);
    }
    /**
     * @param {Object} data
     * @param {Function} [callback]
     * @return {void}
     */

  }, {
    key: "loadInstrumentsWithAddress",
    value: function loadInstrumentsWithAddress(data, callback) {
      var url = this.urlHelper.getTrustedShippingAddressUrl(data.storeId, data.customerId, data.currencyCode);
      var payload = (0, _mappers.mapToTrustedShippingAddressPayload)(data);
      var options = {
        method: _methodTypes.POST,
        headers: (0, _mappers.mapToHeaders)(data)
      };
      this.requestSender.postRequest(url, payload, options, callback);
    }
    /**
     * @param {Object} data
     * @param {Function} [callback]
     * @return {void}
     */

  }, {
    key: "deleteShopperInstrument",
    value: function deleteShopperInstrument(data, callback) {
      var url = this.urlHelper.getInstrumentByIdUrl(data.storeId, data.customerId, data.instrumentId, data.currencyCode);
      var options = {
        method: _methodTypes.DELETE,
        headers: (0, _mappers.mapToHeaders)(data)
      };
      this.requestSender.sendRequest(url, null, options, callback);
    }
  }]);

  return StoreRequestSender;
}();

exports.default = StoreRequestSender;
