// @bigcommerce/bigpay-client/lib/store/url-helper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var UrlHelper = /*#__PURE__*/function () {
  _createClass(UrlHelper, null, [{
    key: "create",

    /**
     * @param {Object} config
     * @param {string} config.host
     * @returns {CustomerMapper}
     */
    value: function create(config) {
      return new UrlHelper(config);
    }
    /**
     * @param {Object} config
     * @param {string} config.host
     * @returns {void}
     */

  }]);

  function UrlHelper(config) {
    _classCallCheck(this, UrlHelper);

    /**
     * @private
     * @type {Object}
     */
    this.config = config;
  }
  /**
   * @private
   * @returns {string}
   * @throws {Error}
   */


  _createClass(UrlHelper, [{
    key: "getInstrumentsUrl",

    /**
     * @param {number} storeId
     * @param {number} customerId
     * @param {string} currencyCode
     * @returns {string}
     */
    value: function getInstrumentsUrl(storeId, customerId, currencyCode) {
      return "".concat(this.host, "/api/v3/stores/").concat(storeId, "/shoppers/").concat(customerId, "/instruments?currency_code=").concat(currencyCode);
    }
    /**
     * @param {number} storeId
     * @param {number} customerId
     * @param {string} currencyCode
     * @return {string}
     */

  }, {
    key: "getTrustedShippingAddressUrl",
    value: function getTrustedShippingAddressUrl(storeId, customerId, currencyCode) {
      return "".concat(this.host, "/api/v3/stores/").concat(storeId, "/shoppers/").concat(customerId, "/instruments/trusted_shipping_address?currency_code=").concat(currencyCode);
    }
    /**
     * @param {number} storeId
     * @param {number} customerId
     * @param {number} instrumentId
     * @param {string} currencyCode
     * @returns {string}
     */

  }, {
    key: "getInstrumentByIdUrl",
    value: function getInstrumentByIdUrl(storeId, customerId, instrumentId, currencyCode) {
      return "".concat(this.host, "/api/v2/stores/").concat(storeId, "/shoppers/").concat(customerId, "/instruments/").concat(instrumentId, "?currency_code=").concat(currencyCode);
    }
  }, {
    key: "host",
    get: function get() {
      if (!this.config || !this.config.host) {
        throw new Error('Host URL unavailable or not supplied.');
      }

      return this.config.host;
    }
  }]);

  return UrlHelper;
}();

exports.default = UrlHelper;
