// @bigcommerce/bigpay-client/lib/client/client.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectAssign = _interopRequireDefault(require("object-assign"));

var _offsitePaymentInitializer = _interopRequireDefault(require("../payment/offsite-payment-initializer"));

var _paymentSubmitter = _interopRequireDefault(require("../payment/payment-submitter"));

var _clientTokenGenerator = _interopRequireDefault(require("../payment/client-token-generator"));

var _storeRequestSender = _interopRequireDefault(require("../store/store-request-sender"));

var _defaultConfig = _interopRequireDefault(require("./default-config"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var Client = /*#__PURE__*/function () {
  _createClass(Client, null, [{
    key: "create",

    /**
     * @param {Object} config
     * @returns {Client}
     */
    value: function create(config) {
      var clientConfig = (0, _objectAssign.default)({}, _defaultConfig.default, config);

      var offsitePaymentInitializer = _offsitePaymentInitializer.default.create(clientConfig);

      var paymentSubmitter = _paymentSubmitter.default.create(clientConfig);

      var clientTokenGenerator = _clientTokenGenerator.default.create(clientConfig);

      var storeRequestSender = _storeRequestSender.default.create(clientConfig);

      return new Client(clientConfig, paymentSubmitter, offsitePaymentInitializer, clientTokenGenerator, storeRequestSender);
    }
    /**
     * @param {Object} config
     * @param {PaymentSubmitter} paymentSubmitter
     * @param {OffsitePaymentInitializer} offsitePaymentInitializer
     * @param {ClientTokenGenerator} clientTokenGenerator
     * @param {StoreRequestSender} storeRequestSender
     */

  }]);

  function Client(config, paymentSubmitter, offsitePaymentInitializer, clientTokenGenerator, storeRequestSender) {
    _classCallCheck(this, Client);

    /**
     * @private
     * @type {Object}
     */
    this.config = config;
    /**
     * @private
     * @type {PaymentSubmitter}
     */

    this.paymentSubmitter = paymentSubmitter;
    /**
     * @private
     * @type {OffsitePaymentInitializer}
     */

    this.offsitePaymentInitializer = offsitePaymentInitializer;
    /**
     * @private
     * @type {ClientTokenGenerator}
     */

    this.clientTokenGenerator = clientTokenGenerator;
    /**
     * @private
     * @type {StoreRequestSender}
     */

    this.storeRequestSender = storeRequestSender;
  }
  /**
   * @param {string} host
   * @returns {void}
   */


  _createClass(Client, [{
    key: "setHost",
    value: function setHost(host) {
      this.config.host = host;
    }
    /**
     * @param {PaymentRequestData} data
     * @param {Function} [callback]
     * @param {string} target
     * @returns {void}
     */

  }, {
    key: "initializeOffsitePayment",
    value: function initializeOffsitePayment(data, callback, target) {
      this.offsitePaymentInitializer.initializeOffsitePayment(data, callback, target);
    }
    /**
     * @param {PaymentRequestData} data
     * @param {Function} [callback]
     * @returns {void}
     */

  }, {
    key: "submitPayment",
    value: function submitPayment(data, callback) {
      this.paymentSubmitter.submitPayment(data, callback);
    }
    /**
     * @param {PaymentRequestData} data
     * @param {Function} [callback]
     * @returns {void}
     */

  }, {
    key: "generateClientToken",
    value: function generateClientToken(data, callback) {
      this.clientTokenGenerator.generateClientToken(data, callback);
    }
    /**
     * @param {Object} data
     * @param {string} data.storeId
     * @param {string} data.customerId
     * @param {string} data.currencyCode
     * @param {Function} [callback]
     * @return {void}
     */

  }, {
    key: "loadInstruments",
    value: function loadInstruments(data, callback) {
      this.storeRequestSender.loadInstruments(data, callback);
    }
    /**
     * @param {Object} data
     * @param {string} data.storeId
     * @param {string} data.customerId
     * @param {string} data.currencyCode
     * @param {AddressData} data.shippingAddress
     * @param {Function} [callback]
     * @return {void}
     */

  }, {
    key: "loadInstrumentsWithAddress",
    value: function loadInstrumentsWithAddress(data, callback) {
      this.storeRequestSender.loadInstrumentsWithAddress(data, callback);
    }
    /**
     * @param {Object} data
     * @param {string} data.storeId
     * @param {string} data.customerId
     * @param {string} data.instrumentId
     * @param {string} data.currencyCode
     * @param {Function} [callback]
     * @return {void}
     */

  }, {
    key: "deleteShopperInstrument",
    value: function deleteShopperInstrument(data, callback) {
      this.storeRequestSender.deleteShopperInstrument(data, callback);
    }
  }]);

  return Client;
}();

exports.default = Client;
