// @bigcommerce/bigpay-client/lib/payment/v1/payment-mappers/customer-mapper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _utils = require("../../../common/utils");

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var CustomerMapper = /*#__PURE__*/function () {
  function CustomerMapper() {
    _classCallCheck(this, CustomerMapper);
  }

  _createClass(CustomerMapper, [{
    key: "mapToCustomer",

    /**
     * @param {PaymentRequestData} data
     * @returns {Object}
     */
    value: function mapToCustomer(data) {
      var _data$customer = data.customer,
          customer = _data$customer === void 0 ? {} : _data$customer,
          _data$quoteMeta = data.quoteMeta,
          quoteMeta = _data$quoteMeta === void 0 ? {} : _data$quoteMeta;
      return (0, _utils.omitNil)({
        customer_group: customer.customerGroupName ? {
          name: customer.customerGroupName
        } : null,
        geo_ip_country_code: quoteMeta.request ? quoteMeta.request.geoCountryCode : null,
        id: customer.customerId ? (0, _utils.toString)(customer.customerId) : null,
        session_token: quoteMeta.request ? quoteMeta.request.sessionHash : null
      });
    }
  }], [{
    key: "create",

    /**
     * @returns {CustomerMapper}
     */
    value: function create() {
      return new CustomerMapper();
    }
  }]);

  return CustomerMapper;
}();

exports.default = CustomerMapper;
