// @bigcommerce/bigpay-client/lib/payment/v1/payment-mappers/payload-mapper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _utils = require("../../../common/utils");

var _customerMapper = _interopRequireDefault(require("./customer-mapper"));

var _orderMapper = _interopRequireDefault(require("./order-mapper"));

var _paymentMapper = _interopRequireDefault(require("./payment-mapper"));

var _storeMapper = _interopRequireDefault(require("./store-mapper"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var PayloadMapper = /*#__PURE__*/function () {
  _createClass(PayloadMapper, null, [{
    key: "create",

    /**
     * @returns {PayloadMapper}
     */
    value: function create() {
      var customerMapper = _customerMapper.default.create();

      var orderMapper = _orderMapper.default.create();

      var paymentMapper = _paymentMapper.default.create();

      var storeMapper = _storeMapper.default.create();

      return new PayloadMapper(customerMapper, orderMapper, paymentMapper, storeMapper);
    }
    /**
     * @param {CustomerMapper} customerMapper
     * @param {OrderMapper} orderMapper
     * @param {PaymentMapper} paymentMapper
     * @param {StoreMapper} storeMapper
     */

  }]);

  function PayloadMapper(customerMapper, orderMapper, paymentMapper, storeMapper) {
    _classCallCheck(this, PayloadMapper);

    /**
     * @private
     * @type {CustomerMapper}
     */
    this.customerMapper = customerMapper;
    /**
     * @private
     * @type {OrderMapper}
     */

    this.orderMapper = orderMapper;
    /**
     * @private
     * @type {PaymentMapper}
     */

    this.paymentMapper = paymentMapper;
    /**
     * @private
     * @type {StoreMapper}
     */

    this.storeMapper = storeMapper;
  }
  /**
   * @param {PaymentRequestData} data
   * @returns {Object}
   */


  _createClass(PayloadMapper, [{
    key: "mapToPayload",
    value: function mapToPayload(data) {
      var _data$order = data.order,
          order = _data$order === void 0 ? {} : _data$order,
          additionalAction = data.additionalAction;
      return (0, _utils.omitNil)({
        customer: this.customerMapper.mapToCustomer(data),
        notify_url: order.callbackUrl,
        order: this.orderMapper.mapToOrder(data),
        payment: this.paymentMapper.mapToPayment(data),
        store: this.storeMapper.mapToStore(data),
        additional_action: additionalAction
      });
    }
    /**
     * @param {PaymentRequestData} data
     * @returns {Object}
     */

  }, {
    key: "mapToHeaders",
    value: function mapToHeaders(data) {
      var authToken = data.authToken;
      return (0, _utils.omitNil)({
        Authorization: authToken
      });
    }
  }]);

  return PayloadMapper;
}();

exports.default = PayloadMapper;
