// @bigcommerce/bigpay-client/lib/payment/client-token-generator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _requestSender = _interopRequireDefault(require("../common/http-request/request-sender"));

var _clientTokenMapper = _interopRequireDefault(require("./v2/payment-mappers/client-token-mapper"));

var _urlHelper = _interopRequireDefault(require("./url-helper"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var ClientTokenGenerator = /*#__PURE__*/function () {
  _createClass(ClientTokenGenerator, null, [{
    key: "create",

    /**
     * @param {Object} config
     * @returns {PaymentSubmitter}
     */
    value: function create(config) {
      var urlHelper = _urlHelper.default.create(config);

      var requestSender = _requestSender.default.create();

      var clientTokenMapper = _clientTokenMapper.default.create();

      return new ClientTokenGenerator(urlHelper, requestSender, clientTokenMapper);
    }
    /**
     * @param {UrlHelper} urlHelper
     * @param {RequestSender} requestSender
     * @param {ClientTokenMapper} clientTokenMapper
     * @returns {void}
     */

  }]);

  function ClientTokenGenerator(urlHelper, requestSender, clientTokenMapper) {
    _classCallCheck(this, ClientTokenGenerator);

    /**
     * @private
     * @type {UrlHelper}
     */
    this.urlHelper = urlHelper;
    /**
     * @private
     * @type {RequestSender}
     */

    this.requestSender = requestSender;
    /**
     * @private
     * @type {ClientTokenMapper}
     */

    this.clientTokenMapper = clientTokenMapper;
  }
  /**
   * @param {PaymentRequestData} data
   * @param {Function} [callback]
   * @returns {void}
   */


  _createClass(ClientTokenGenerator, [{
    key: "generateClientToken",
    value: function generateClientToken(data, callback) {
      var url = this.urlHelper.getGenerateClientTokenUrl();
      var payload = this.clientTokenMapper.mapToClientToken(data);
      this.requestSender.postRequest(url, payload, {}, callback);
    }
  }]);

  return ClientTokenGenerator;
}();

exports.default = ClientTokenGenerator;
