// @bigcommerce/bigpay-client/lib/payment/offsite-payment-initializer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _formPoster = require("@bigcommerce/form-poster");

var _paymentTypes = require("./payment-types");

var _payloadMapper = _interopRequireDefault(require("./offsite-payment-mappers/payload-mapper"));

var _urlHelper = _interopRequireDefault(require("./url-helper"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var OffsitePaymentInitializer = /*#__PURE__*/function () {
  _createClass(OffsitePaymentInitializer, null, [{
    key: "create",

    /**
     * @param {Object} config
     * @returns {OffsitePaymentInitializer}
     */
    value: function create(config) {
      var urlHelper = _urlHelper.default.create(config);

      var formPoster = (0, _formPoster.createFormPoster)();

      var payloadMapper = _payloadMapper.default.create();

      return new OffsitePaymentInitializer(urlHelper, formPoster, payloadMapper);
    }
    /**
     * @param {UrlHelper} urlHelper
     * @param {FormPoster} formPoster
     * @param {PayloadMapper} payloadMapper
     * @returns {void}
     */

  }]);

  function OffsitePaymentInitializer(urlHelper, formPoster, payloadMapper) {
    _classCallCheck(this, OffsitePaymentInitializer);

    /**
     * @private
     * @type {UrlHelper}
     */
    this.urlHelper = urlHelper;
    /**
     * @private
     * @type {FormPoster}
     */

    this.formPoster = formPoster;
    /**
     * @private
     * @type {PayloadMapper}
     */

    this.payloadMapper = payloadMapper;
  }
  /**
   * @param {PaymentRequestData} data
   * @param {Function} [callback]
   * @param {string} target
   * @returns {void}
   * @throws {Error}
   */


  _createClass(OffsitePaymentInitializer, [{
    key: "initializeOffsitePayment",
    value: function initializeOffsitePayment(data, callback, target) {
      var _data$paymentMethod = data.paymentMethod,
          paymentMethod = _data$paymentMethod === void 0 ? {} : _data$paymentMethod;

      if (paymentMethod.type !== _paymentTypes.HOSTED) {
        throw new Error("".concat(paymentMethod.type, " is not supported."));
      }

      var payload = this.payloadMapper.mapToPayload(data);
      var url = this.urlHelper.getOffsitePaymentUrl();
      this.formPoster.postForm(url, payload, callback, target);
    }
  }]);

  return OffsitePaymentInitializer;
}();

exports.default = OffsitePaymentInitializer;
