// @bigcommerce/bigpay-client/lib/payment/payment-method-ids.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PAYPAL_COMMERCE_CREDIT_CARDS = exports.PAYPAL_COMMERCE_CREDIT = exports.PAYPAL_COMMERCE_ALTERNATIVE_METHODS = exports.PAYPAL_COMMERCE = exports.BRAINTREE_GOOGLEPAY = exports.BRAINTREE_VISACHECKOUT = exports.BRAINTREE_PAYPAL_CREDIT = exports.BRAINTREE_PAYPAL = exports.BRAINTREE = void 0;
var BRAINTREE = 'braintree';
exports.BRAINTREE = BRAINTREE;
var BRAINTREE_PAYPAL = 'braintreepaypal';
exports.BRAINTREE_PAYPAL = BRAINTREE_PAYPAL;
var BRAINTREE_PAYPAL_CREDIT = 'braintreepaypalcredit';
exports.BRAINTREE_PAYPAL_CREDIT = BRAINTREE_PAYPAL_CREDIT;
var BRAINTREE_VISACHECKOUT = 'braintreevisacheckout';
exports.BRAINTREE_VISACHECKOUT = BRAINTREE_VISACHECKOUT;
var BRAINTREE_GOOGLEPAY = 'googlepaybraintree';
exports.BRAINTREE_GOOGLEPAY = BRAINTREE_GOOGLEPAY;
var PAYPAL_COMMERCE = 'paypalcommerce';
exports.PAYPAL_COMMERCE = PAYPAL_COMMERCE;
var PAYPAL_COMMERCE_ALTERNATIVE_METHODS = 'paypalcommercealternativemethods';
exports.PAYPAL_COMMERCE_ALTERNATIVE_METHODS = PAYPAL_COMMERCE_ALTERNATIVE_METHODS;
var PAYPAL_COMMERCE_CREDIT = 'paypalcommercecredit';
exports.PAYPAL_COMMERCE_CREDIT = PAYPAL_COMMERCE_CREDIT;
var PAYPAL_COMMERCE_CREDIT_CARDS = 'paypalcommercecreditcards';
exports.PAYPAL_COMMERCE_CREDIT_CARDS = PAYPAL_COMMERCE_CREDIT_CARDS;
