// @bigcommerce/bigpay-client/lib/payment/v2/payment-mappers/client-token-mapper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _utils = require("../../../common/utils");

var _cartMapper = _interopRequireDefault(require("./cart-mapper"));

var _gatewayMapper = _interopRequireDefault(require("./gateway-mapper"));

var _quoteMapper = _interopRequireDefault(require("./quote-mapper"));

var _storeMapper = _interopRequireDefault(require("./store-mapper"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var ClientTokenMapper = /*#__PURE__*/function () {
  _createClass(ClientTokenMapper, null, [{
    key: "create",

    /**
     * @returns {ClientTokenMapper}
     */
    value: function create() {
      var cartMapper = _cartMapper.default.create();

      var gatewayMapper = _gatewayMapper.default.create();

      var quoteMapper = _quoteMapper.default.create();

      var storeMapper = _storeMapper.default.create();

      return new ClientTokenMapper(cartMapper, gatewayMapper, quoteMapper, storeMapper);
    }
    /**
     * @param {CartMapper} cartMapper
     * @param {GatewayMapper} gatewayMapper
     * @param {QuoteMapper} quoteMapper
     * @param {StoreMapper} storeMapper
     */

  }]);

  function ClientTokenMapper(cartMapper, gatewayMapper, quoteMapper, storeMapper) {
    _classCallCheck(this, ClientTokenMapper);

    /**
     * @private
     * @type {CartMapper}
     */
    this.cartMapper = cartMapper;
    /**
     * @private
     * @type {GatewayMapper}
     */

    this.gatewayMapper = gatewayMapper;
    /**
     * @private
     * @type {QuoteMapper}
     */

    this.quoteMapper = quoteMapper;
    /**
     * @private
     * @type {StoreMapper}
     */

    this.storeMapper = storeMapper;
  }
  /**
   * @param {PaymentRequestData} data
   * @returns {Object}
   */


  _createClass(ClientTokenMapper, [{
    key: "mapToClientToken",
    value: function mapToClientToken(data) {
      return (0, _utils.omitNil)({
        cart: this.cartMapper.mapToCart(data),
        gateway: this.gatewayMapper.mapToGateway(data),
        quote: this.quoteMapper.mapToQuote(data),
        store: this.storeMapper.mapToStore(data)
      });
    }
  }]);

  return ClientTokenMapper;
}();

exports.default = ClientTokenMapper;
