// @bigcommerce/bigpay-client/lib/common/http-request/request-sender.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _deepAssign = _interopRequireDefault(require("deep-assign"));

var _objectAssign = _interopRequireDefault(require("object-assign"));

var _methodTypes = require("./method-types");

var _defaultOptions = _interopRequireDefault(require("./default-options"));

var _payloadTransformer = _interopRequireDefault(require("./payload-transformer"));

var _requestFactory = _interopRequireDefault(require("./request-factory"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var RequestSender = /*#__PURE__*/function () {
  _createClass(RequestSender, null, [{
    key: "create",

    /**
     * @returns {RequestSender}
     */
    value: function create() {
      var requestFactory = _requestFactory.default.create();

      var payloadTransformer = _payloadTransformer.default.create();

      return new RequestSender(requestFactory, payloadTransformer);
    }
    /**
     * @param {RequestFactory} requestFactory
     * @param {PayloadTransformer} payloadTransformer
     */

  }]);

  function RequestSender(requestFactory, payloadTransformer) {
    _classCallCheck(this, RequestSender);

    /**
     * @private
     * @type {RequestFactory}
     */
    this.requestFactory = requestFactory;
    /**
     * @private
     * @type {PayloadTransformer}
     */

    this.payloadTransformer = payloadTransformer;
  }
  /**
   * @param {string} url
   * @param {Object} data
   * @param {Object} [options]
   * @param {Function} [callback]
   * @returns {void}
   */


  _createClass(RequestSender, [{
    key: "sendRequest",
    value: function sendRequest(url, data, options, callback) {
      var _this = this;

      var mergedOptions = (0, _deepAssign.default)({}, _defaultOptions.default, options);
      var xhr = this.requestFactory.createRequest(url, mergedOptions, function (error) {
        var response = _this.payloadTransformer.fromResponse(xhr);

        if (!callback) {
          return;
        }

        if (error || !_this.isSuccessfulRequest(xhr)) {
          callback(response);
          return;
        }

        callback(null, response);
      });
      var payload = this.payloadTransformer.toRequest(data, mergedOptions.headers['Content-Type']);
      xhr.send(payload);
    }
    /**
     * @param {string} url
     * @param {Object} data
     * @param {Object} [options]
     * @param {Function} [callback]
     * @returns {void}
     */

  }, {
    key: "postRequest",
    value: function postRequest(url, data, options, callback) {
      var mergedOptions = (0, _objectAssign.default)({}, options, {
        method: _methodTypes.POST
      });
      this.sendRequest(url, data, mergedOptions, callback);
    }
    /**
     * @private
     * @param {XMLHttpRequest} xhr
     * @returns {boolean}
     */

  }, {
    key: "isSuccessfulRequest",
    value: function isSuccessfulRequest(xhr) {
      return xhr.status >= 200 && xhr.status < 300;
    }
  }]);

  return RequestSender;
}();

exports.default = RequestSender;
