// @bigcommerce/bigpay-client/lib/payment/offsite-payment-mappers/customer-mapper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _utils = require("../../common/utils");

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var CustomerMapper = /*#__PURE__*/function () {
  function CustomerMapper() {
    _classCallCheck(this, CustomerMapper);
  }

  _createClass(CustomerMapper, [{
    key: "mapToCustomer",

    /**
     * @param {PaymentRequestData} data
     * @returns {Object}
     */
    value: function mapToCustomer(data) {
      var _data$customer = data.customer,
          customer = _data$customer === void 0 ? {} : _data$customer,
          _data$quoteMeta = data.quoteMeta,
          quoteMeta = _data$quoteMeta === void 0 ? {} : _data$quoteMeta,
          _data$store = data.store,
          store = _data$store === void 0 ? {} : _data$store;
      return (0, _utils.omitNil)({
        customer_browser_info: navigator.userAgent,
        customer_email: customer.email,
        customer_first_name: customer.firstName,
        customer_geo_ip_country_code: quoteMeta.request ? quoteMeta.request.geoCountryCode : null,
        customer_id: customer.customerId,
        customer_last_name: customer.lastName,
        customer_locale: store.storeLanguage,
        customer_name: customer.name,
        customer_phone: customer.phoneNumber,
        customer_reference: customer.email
      });
    }
  }], [{
    key: "create",

    /**
     * @returns {CustomerMapper}
     */
    value: function create() {
      return new CustomerMapper();
    }
  }]);

  return CustomerMapper;
}();

exports.default = CustomerMapper;
