// @bigcommerce/bigpay-client/lib/payment/offsite-payment-mappers/meta-mapper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _utils = require("../../common/utils");

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var MetaMapper = /*#__PURE__*/function () {
  function MetaMapper() {
    _classCallCheck(this, MetaMapper);
  }

  _createClass(MetaMapper, [{
    key: "mapToMeta",

    /**
     * @param {PaymentRequestData} data
     * @returns {Object}
     */
    value: function mapToMeta(data) {
      var source = data.source;
      return (0, _utils.omitNil)({
        meta_referrer: document.referrer,
        meta_source: source,
        meta_user_agent: navigator.userAgent
      });
    }
  }], [{
    key: "create",

    /**
     * @returns {MetaMapper}
     */
    value: function create() {
      return new MetaMapper();
    }
  }]);

  return MetaMapper;
}();

exports.default = MetaMapper;
